### PR TITLE
Make extension discriminator clear

### PR DESCRIPTION
**Pull request by Bakugo90**
Implement #4
For this pull request i work on "A" and "M" extension. 
I've cloned your JSON branch and worked from that so I can continue what you've already done.

> I have already implemented all of the necessary changes. Look at the "JSON" branch in this repository. Further work here should probably be on top of that branch.

Of course, I've noticed that you've already worked on the "V" and "Zbkb" extension. And that really helped me to recognise the "M" and "A" extensions.

### CODE REVIEW

**Comment by ThinkOpenly** [comment]
Greetings, and thanks for this submission!

While the changes you have implemented are not incorrect, they do not address the issue described in #4.

If you look at most of the changes in the existing code where the `extension` function is used, you'll see it mostly is appended to (one or both sides of) `mapping clause encdec` definitions. Issue #4 is asking that the same type of conditional expressions used for some extensions be used for _all_ extensions. For example, the current code has:
```
mapping clause encdec = LOADRES(aq, rl, rs1, size, rd) if amo_width_valid(size)
```
and it would need to be something like:
```
mapping clause encdec = LOADRES(aq, rl, rs1, size, rd) if extension("A") & amo_width_valid(size)
```

File: model/riscv_insts_mext.sail [hunk 1]
```diff
@@ -148,11 +148,11 @@ mapping clause assembly = DIV(rs2, rs1, rd, s)
 /* ****************************************************************** */
 union clause ast = REM : (regidx, regidx, regidx, bool)
 
-mapping clause encdec = REM(rs2, rs1, rd, s)
+mapping clause encdec = REM(rs2, rs1, rd, s) & extension("M")
```
Comment (ThinkOpenly) [inline model/riscv_insts_mext.sail:151]: The `&` should be `if`.

File: model/riscv_insts_mext.sail [hunk 2]
```diff
@@ -173,13 +173,13 @@ mapping clause assembly = REM(rs2, rs1, rd, s)
 /* ****************************************************************** */
 union clause ast = MULW : (regidx, regidx, regidx)
 
-mapping clause encdec = MULW(rs2, rs1, rd)
+mapping clause encdec = MULW(rs2, rs1, rd) if extension("M")
       if sizeof(xlen) == 64
   <-> 0b0000001 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0111011
       if sizeof(xlen) == 64
```
Comment (ThinkOpenly) [inline model/riscv_insts_mext.sail:179]: This doesn't compile. You need to combine the new condition with the existing condition, and on both sides of the mapping, like this:
```
mapping clause encdec = MULW(rs2, rs1, rd)
      if extension("M") & sizeof(xlen) == 64
  <-> 0b0000001 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0111011
      if extension("M") & sizeof(xlen) == 64
```

File: model/riscv_insts_mext.sail [hunk 3]
```diff
@@ -203,13 +203,13 @@ mapping clause assembly = MULW(rs2, rs1, rd)
 /* ****************************************************************** */
 union clause ast = DIVW : (regidx, regidx, regidx, bool)
 
-mapping clause encdec = DIVW(rs2, rs1, rd, s)
+mapping clause encdec = DIVW(rs2, rs1, rd, s) if extension("M")
```
Comment (ThinkOpenly) [inline model/riscv_insts_mext.sail:206]: same as previous comment.

File: model/riscv_insts_mext.sail [hunk 4]
```diff
@@ -233,13 +233,13 @@ mapping clause assembly = DIVW(rs2, rs1, rd, s)
 /* ****************************************************************** */
 union clause ast = REMW : (regidx, regidx, regidx, bool)
 
-mapping clause encdec = REMW(rs2, rs1, rd, s)
+mapping clause encdec = REMW(rs2, rs1, rd, s) if extension("M")
```
Comment (ThinkOpenly) [inline model/riscv_insts_mext.sail:236]: same as previous comment.

**Review by ThinkOpenly — CHANGES_REQUESTED** [review]
A few issues to fix. If you want to fix this up and resubmit, that would be ideal. If you prefer that I fix them up and merge, let me know.

**Comment by Bakugo90** [comment]
Done ! @ThinkOpenly

**Review by ThinkOpenly — APPROVED** [review]

**Comment by ThinkOpenly** [comment]
FYI, this commit and other related "extension tagging" commits have been merged upstream. https://github.com/riscv/sail-riscv/commit/ebf57cc79459bdba3e8d15e84941b6dcd088f5f4. Congratulations, and thank you for your efforts! :tada:

**Pull request merged**